lib/http2/stream.c

```diff
@@ -190,10 +193,10 @@ static h2o_iovec_t *send_data_push(h2o_http2_conn_t *conn, h2o_http2_stream_t *s
     /* commit the DATA frame if we have actually emitted payload */
     if (dst.len != max_payload_size || !h2o_send_state_is_in_progress(send_state)) {
         size_t payload_len = max_payload_size - dst.len;
-        encode_data_header_and_consume_window(conn, stream, (uint8_t *)conn->_write.buf->bytes + conn->_write.buf->size,
-                                              payload_len, send_state,
-                                              stream->state >= H2O_HTTP2_STREAM_STATE_SEND_BODY_IS_FINAL && bufcnt == 0);
-        conn->_write.buf->size += H2O_HTTP2_FRAME_HEADER_SIZE + payload_len;
+        if (bufcnt != 0 && send_state == H2O_SEND_STATE_FINAL) {
```

Comment:
> `send_state == H2O_SEND_STATE_FINAL`

Do we need this check?

Please correct me if I'm wrong, bit to me it seems that we should not let `encode_data_header_and_consume_window` send neither `END_STREAM` or `RST_STREAM` until all pending data gets sent (i.e. when `bufcnt` becomes zero and send state is not in-progress).